### PR TITLE
refactor(ci): pull git context out to separate action

### PR DESCRIPTION
**Pull request by rwaskiewicz**
## Dependencies

This PR depends on #3626 landing first

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Build (`npm run build`) was run locally and any changes were pushed
- [ ] Unit tests (`npm test`) were run locally and passed
- [ ] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [x] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Running some tests in Stencil can generated a new `components.d.ts` file
when making modications to the compiler. Currently, we only programmatically
check for this in our BrowserStack based tests, which leads to a lot of "by hand"
checking for un-checked-in/updated files (or, worse, we completely miss them)

GitHub Issue Number: N/A

## What is the new behavior?

this commit pulls the check of the git context out to a separate,
reusable action. this allows us to use it in other places than only
the browserstack job. it has been added to all of our test suites,
which also may generate a new `components.d.ts` file as a result
of changes to the compiler, then building the tests.

a reusable action was created here instead of a reusable _job_
(that would run upon completion of all tests that are run during ci)
due to the design of our ci. currently, each reusable job must check
out its own version of the codebase. that makes it tricky to check whether
or not we have a dirty git context if you have to check out the code.
rather than attempt anything more complicated here, the trade-off
of a little more duplication in favor of a working solution for this out
of band work felt like the right path

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

I originally called out an area of the codebase that I felt might have unchecked in changes in https://github.com/ionic-team/stencil/pull/3612#discussion_r975338991. I cherry-picked the commits from that PR that caused the `components.d.ts` files to change ([8f433d4](https://github.com/ionic-team/stencil/pull/3625/commits/8f433d4535343ba17bd09597e298c408fa78a90e), [8f433d4](https://github.com/ionic-team/stencil/pull/3625/commits/8f433d4535343ba17bd09597e298c408fa78a90e)) and was able to observe:
- Browserstack failing ([CI logs](https://github.com/ionic-team/stencil/actions/runs/3091448758/jobs/5001600705))
- Analysis + E2E tests failing ([CI logs](https://github.com/ionic-team/stencil/actions/runs/3091448741/jobs/5001688607))

I then reverted those cherry-picked commits to see everything passing again

## Other information

The analysis test is also building the e2e tests:

> stencil-test@ build /home/runner/work/stencil/stencil/test
> npm run build.browser-compile && npm run build.hello-world && npm run build.hello-vdom && npm run build.todo && npm run build.end-to-end && npm run build.ionic

Something we can do as a quick win to speed things up once we've decided what to do with them (the analysis tests)

### CODE REVIEW

File: .github/workflows/actions/check-git-context/action.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+name: 'Check Git Context'
+description: 'checks for a dirty git context, failing if the context is dirty'
+runs:
+  using: composite
+  steps:
+    - name: Git status check
+      # here we check that there are no changed / new files.
+      # we use `git status`, grep out the build zip used throughout CI,
+      # and check if there are more than 0 lines in the output.
```
Comment (alicewriteswrongs) [inline .github/workflows/actions/check-git-context/action.yml:9]: maybe we should just update this comment real quick to say

```suggestion
      # here we check that there are no changed / new files.
      # we use `git status`, grep out the build zip used throughout CI,
      # and check if there are more than 0 lines in the output.
      # if there are more than 0 lines (which indicates new files written
      # during the CI run) we print out the return value of `git status`
```

or something along those lines, just to accurately describe what we do here

**Review by alicewriteswrongs — CHANGES_REQUESTED** [review]
just had one question / suggestion

File: .github/workflows/actions/check-git-context/action.yml (hunk 1 above)
Comment (rwaskiewicz) [reply .github/workflows/actions/check-git-context/action.yml:9]: 👍 I dig it - done in https://github.com/ionic-team/stencil/pull/3625/commits/21d0c85b55a2549b408aae182af2808fdcea1a65

**Review by alicewriteswrongs — APPROVED** [review]
👍 lgtm!

**Review by tanner-reits — APPROVED** [review]

**Pull request merged**